packages/medusa/src/services/payment-provider.ts

```diff
@@ -696,4 +747,114 @@ export default class PaymentProviderService extends TransactionBaseService {
+    const cart =
+      "object" in cartOrData && cartOrData.object === "cart"
+        ? cartOrData
+        : ((cartOrData as PaymentSessionInput).cart as Cart)
+
+    const context = {} as Cart & PaymentContext
+
+    // TODO: only to support legacy API. Once we are ready to break the API, the cartOrData will only support PaymentSessionInput
+    if ("object" in cartOrData && cartOrData.object === "cart") {
+      context.cart = {
+        context: cart.context,
+        shipping_address: cart.shipping_address,
+        id: cart.id,
+        email: cart.email,
+        shipping_methods: cart.shipping_methods,
+      }
+      context.amount = cart.total!
+      context.currency_code = cart.region?.currency_code
+      context.collected_data = cart.customer?.metadata ?? {}
+      Object.assign(context, cart)
+    } else {
+      const data = cartOrData as PaymentSessionInput
+      context.cart = data.cart
+      context.amount = data.amount
+      context.currency_code = data.currency_code
+      context.collected_data = data.customer?.metadata ?? {}
+      Object.assign(context, cart)
+    }
+
+    return context
+  }
+
+  /**
+   * Persist a Payment session data
+   * @param providerId
+   * @param data
+   * @protected
+   */
+  protected async saveSession(
+    providerId: string,
+    data: {
+      cartId?: string
+      amount?: number
+      sessionData: Record<string, unknown>
+      isSelected?: boolean
+      status: PaymentSessionStatus
+    }
+  ): Promise<PaymentSession> {
+    const manager = this.transactionManager_ ?? this.manager_
+
+    if (
+      data.amount != null &&
+      !this.featureFlagRouter_.isFeatureEnabled(OrderEditingFeatureFlag.key)
+    ) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_ARGUMENT,
+        "Unable to save the payment session with an amount. The feature flag order edit is not enabled."
+      )
+    }
+
+    const sessionRepo = manager.getCustomRepository(
+      this.paymentSessionRepository_
+    )
+
+    const toCreate: Partial<PaymentSession> = {
+      cart_id: data.cartId,
+      provider_id: providerId,
+      data: data.sessionData,
+      is_selected: data.isSelected,
+      status: data.status,
+      amount: data.amount,
+    }
+
+    const created = sessionRepo.create(toCreate)
+    return await sessionRepo.save(created)
+  }
+
+  /**
+   * Process the collected data. Can be used every time we need to process some collected data returned by the provider
+   * @param data
+   * @param paymentResponse
+   * @protected
+   */
+  protected async processCollectedData(
+    data: { customer?: { id?: string } } = {},
+    paymentResponse: PaymentSessionResponse | Record<string, unknown>
+  ): Promise<void> {
+    const { collected_data } = paymentResponse as PaymentSessionResponse
+
+    if (!collected_data) {
+      return
+    }
+
+    const manager = this.transactionManager_ ?? this.manager_
+
+    if (collected_data.customer && data.customer?.id) {
```

Comment:
question: Should we call it something else than `customer` when it is collected from the payment provider to avoid ambiguity?

E.g. `provider_customer_id` or something similar

Reply:
This comment might be irrelevant if we choose to handle the collected data differently as per @srindom's comments